src/dg/dg.cpp

```diff
@@ -750,33 +750,36 @@ void DGBase<dim,real,MeshType>::assemble_cell_residual (
     const dealii::FESystem<dim> &fe_metric = this->high_order_grid->fe_system;
     const unsigned int n_metric_dofs = fe_metric.dofs_per_cell;
     const unsigned int n_grid_nodes  = n_metric_dofs / dim;
-    //Rewrite the high_order_grid->volume_nodes in a way we can use sum-factorization on.
-    //That is, splitting up the vector by the dimension.
+
     std::array<std::vector<real>,dim> mapping_support_points;
-    for(int idim=0; idim<dim; idim++){
-        mapping_support_points[idim].resize(n_grid_nodes);
-    }
-    for (unsigned int igrid_node = 0; igrid_node< n_metric_dofs/dim; ++igrid_node) {
-        for (unsigned int idof = 0; idof< n_metric_dofs; ++idof) {
-            const real val = (high_order_grid->volume_nodes[current_metric_dofs_indices[idof]]);
-            const unsigned int istate = fe_metric.system_to_component_index(idof).first; 
-            mapping_support_points[istate][igrid_node] += val * fe_metric.shape_value_component(idof,high_order_grid->dim_grid_nodes.point(igrid_node),istate); 
-        }
-    }
     //if have source term need to store vol flux nodes.
     const bool store_vol_flux_nodes = all_parameters->manufactured_convergence_study_param.manufactured_solution_param.use_manufactured_source_term;
     //for boundary conditions not periodic we need surface flux nodes
     const bool store_surf_flux_nodes = (all_parameters->use_periodic_bc) ? false : true;
-
     OPERATOR::metric_operators<real,dim,2*dim> metric_oper_int(nstate, poly_degree, grid_degree,
                                                                store_vol_flux_nodes,
                                                                store_surf_flux_nodes);
-    //build the volume metric cofactor matrix and the determinant of the volume metric Jacobian
-    metric_oper_int.build_volume_metric_operators(
-        volume_quadrature_collection[poly_degree].size(), n_grid_nodes,
-        mapping_support_points,
-        mapping_basis,
-        this->all_parameters->use_invariant_curl_form);
+    if(use_strong_form){
```

Comment:
Careful about peppering those flags around. Parent class should not behave differently for the various derived classes. Use polymorphism rather than if-statements.

Reply:
I agree. The issue I faced was that, for example, the assemble volume functions take drastically different inputs now for DG strong and DG Weak (they're basically 2 different codes), as well as needing to allocate different terms. ie Weak still uses FEValues, whereas strong uses the operators. That is why I felt for convenience/quick fix, it would be simpler to make a new function assemble_volume_strong and check with a flag.
To make sure I understand correctly for the polymorphism: I should create, per say a function, solve_volume, where it would then in DGStrong create the operators and call assemble_volume_term_strong, whereas in DGWeak it would build FEValues and call its volume functions, but in DGBase it doesn't distinguish between either or?
Lastly, would it be ok if I put this as an issue for the future?

Reply:
Yup, with polymorphism, you let the class hierarchy dictate what logic instead of using if-statements. While both WeakDG and StrongDG are both DGBase, they behave differently based on which one is the concrete class being constructed. If you put logic in DGBase, why bother using WeakDG and StrongDG? We could just have a big DGBase class with both Weak and Strong DG in it with some if-statements.

The hiccup about convenience and quick fixes for you usually means inconvenience for others and longer fixes for others. Someone who needs to work the specifics of StrongDG now needs to be aware about something happening somewhere else (I see ~20 other places where this is used). 

Leaving it up to the others to review this PR and balance what they deem worthy of fixing now or later.

Reply:
Fixed. I removed the flag, and it now uses polymorphism for the assemble functions and to build needed operators/fevalues etc on the fly in the cell.